src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,686 @@
+|x|≡0⇒x≡0 {suc2* _} ()
+
+--============================================================================
+-- Arithmetic operations on Bin and their properties.
+
+suc : Bin → Bin
+suc 0#        =  suc2* 0#
+suc (2suc x)  =  suc2* (suc x)   -- 1 + 2(1+x)
+suc (suc2* x) =  2suc x          -- 1 + 1 + 2x =  2*(1+x)
+
+1# = suc 0#;   2# = suc 1#;   3# = suc 2#;   4# = suc 3#;   5# = suc 4#
+
+---------------------------
+suc≢0 :  ∀ {x} → suc x ≢ 0#
+suc≢0 {0#}      ()
+suc≢0 {2suc _}  ()
+suc≢0 {suc2* _} ()
+
+------------
+infixl 6 _+_
+
+_+_ : Op₂ Bin
+0#        + y         =  y
+x         + 0#        =  x
+(2suc x)  + (2suc y)  =  2suc (suc (x + y))
+                                       -- 2(1+x) + 2(1+y) =  2(1 + 1+x+y)
+(2suc x)  + (suc2* y) =  suc (2suc (x + y))
+                                       -- 2(1+x) + 1 + 2y =  1 + 2(1+x+y)
+(suc2* x) + (2suc y)  =  suc (2suc (x + y))
+(suc2* x) + (suc2* y) =  suc (suc2* (x + y))
+                              -- 1+2x + 1+2y =  2 + 2(x+y) =  1 + 1 + 2(x+y)
+
+------------------------------------------------------------
+x≢0⇒x+y≢0 :  {x : Bin} (y : Bin) → x ≢ 0# → x + y ≢ 0#
+x≢0⇒x+y≢0 {2suc _} 0# _    =  λ()
+x≢0⇒x+y≢0 {0#}     _  0≢0 =  contradiction refl 0≢0
+
+---------------
+2* : Bin → Bin
+2* 0#        = 0#
+2* (2suc x)  = 2suc (suc2* x)
+                            -- 2(1+x) + 2(1+x) = 2(1+x + 1+x) =  2(1 + 1+2x)
+2* (suc2* x) = 2suc (2* x)
+
+-----------------
+pred : Bin → Bin
+pred 0#        =  0#
+pred (2suc x)  =  suc2* x   -- 2(1+x) - 1 =  1+2x
+pred (suc2* x) =  2* x      -- 1 + 2x -1  =  2x
+
+---------------------------
+2suc-as∘ :  2suc ≗ 2* ∘ suc
+2suc-as∘ 0#        =  refl
+2suc-as∘ (2suc x)  =  begin 2suc (2suc x)        ≡⟨ cong 2suc (2suc-as∘ x) ⟩
+                            2suc (2* (suc x))    ≡⟨⟩
+                            2* (suc2* (suc x))   ≡⟨⟩
+                            2* (suc (2suc x))
+                      ∎
+2suc-as∘ (suc2* x) =  refl
+
+
+-- suc2*∘suc≗suc∘2suc :  suc2* ∘ suc ≗ suc ∘ 2suc   -- is by  const refl
+
+suc2*-as∘ :  suc2* ≗ suc ∘ 2*
+suc2*-as∘ 0#        = refl
+suc2*-as∘ (2suc x)  = refl
+suc2*-as∘ (suc2* x) =
+  begin suc2* (suc2* x)     ≡⟨ cong suc2* (suc2*-as∘ x) ⟩
+        suc2* (suc 2x)      ≡⟨⟩
+        suc (2suc 2x)       ≡⟨ cong suc (2suc-as∘ 2x) ⟩
+        suc (2* (suc 2x))   ≡⟨ cong (suc ∘ 2*) (sym (suc2*-as∘ x)) ⟩
+        suc (2* (suc2* x))
+  ∎
+  where 2x = 2* x
+
+---------------------------
+pred∘suc :  pred ∘ suc ≗ id
+pred∘suc 0#        =  refl
+pred∘suc (2suc x)  =  sym (2suc-as∘ x)
+pred∘suc (suc2* x) =  refl
+
+
+suc∘pred :  (x : Bin) → x ≢ 0# → suc (pred x) ≡ x
+
+suc∘pred 0#        0≢0 =  contradiction refl 0≢0
+suc∘pred (2suc _)  _   =  refl
+suc∘pred (suc2* x) _   =  begin suc (pred (suc2* x))   ≡⟨⟩
+                                suc (2* x)             ≡⟨ sym (suc2*-as∘ x) ⟩
+                                suc2* x
+                          ∎
+
+------------------------------
+0+ :  (x : Bin) → 0# + x ≡ x
```

Comment:
Please follow standard library naming conventions for these properties

Reply:
Done.
``ε∙, 0+, 0*``  was my proposal, to make proofs more readable/writable.